pkg/util/merger.go

```diff
@@ -0,0 +1,50 @@
+package util
+
+import "github.com/prometheus/common/model"
+
+// MergeSampleSets merges and dedupes two sets of already sorted sample pairs.
+func MergeSampleSets(a, b []model.SamplePair) []model.SamplePair {
+	result := make([]model.SamplePair, 0, len(a)+len(b))
+	i, j := 0, 0
+	for i < len(a) && j < len(b) {
+		if a[i].Timestamp < b[j].Timestamp {
+			result = append(result, a[i])
+			i++
+		} else if a[i].Timestamp > b[j].Timestamp {
+			result = append(result, b[j])
+			j++
+		} else {
+			result = append(result, a[i])
+			i++
+			j++
+		}
+	}
+	for ; i < len(a); i++ {
+		result = append(result, a[i])
+	}
+	for ; j < len(b); j++ {
+		result = append(result, b[j])
+	}
+	return result
+}
+
+// MergeNSampleSets merges and dedupes n sets of already sorted sample pairs.
+func MergeNSampleSets(sampleSets ...[]model.SamplePair) []model.SamplePair {
```

Comment:
Would it be worth implementing: 
- a n-way merge, instead of 
- this 2-way merge for which we then enqueue results (and repeat)?

It feels like we would re-allocate quite a few arrays and reprocess data for large number of `sampleSets`.